Race condition error: listen EACCES: permission denied ::0:25

I am testing latest master and some(!!) machines has serious problem starting haraka at all.

**previous versions:**
```
2024-05-21 14:30:10.573300500  [NOTICE] [-] [core] Listening on [::0]:25
2024-05-21 14:30:10.573718500  [NOTICE] [-] [server] Switching from current gid: 0
2024-05-21 14:30:10.574419500  [NOTICE] [-] [server] New gid: 8
2024-05-21 14:30:10.574470500  [NOTICE] [-] [server] Switching from current uid: 0
2024-05-21 14:30:10.574761500  [NOTICE] [-] [server] New uid: 88
```

**master:**
```
2024-05-21 13:44:41.261384500  [NOTICE] [-] [server] Switching from current gid: 0
2024-05-21 13:44:41.262136500  [NOTICE] [-] [server] New gid: 8
2024-05-21 13:44:41.262237500  [NOTICE] [-] [server] Switching from current uid: 0
2024-05-21 13:44:41.262567500  [NOTICE] [-] [server] New uid: 88
...
2024-05-21 13:44:41.275302500  [ERROR] [-] [server] Failed to setup listeners: listen EACCES: permission denied ::0:25
2024-05-21 13:44:41.278932500  [ERROR] [-] [server] Error: listen EACCES: permission denied ::0:25
2024-05-21 13:44:41.278940500      at Server.setupListenHandle [as _listen2] (node:net:1800:21)
2024-05-21 13:44:41.278944500      at listenInCluster (node:net:1865:12)
2024-05-21 13:44:41.278948500      at doListen (node:net:2014:7)
2024-05-21 13:44:41.278952500      at process.processTicksAndRejections (node:internal/process/task_queues:83:21) {
2024-05-21 13:44:41.278956500    code: 'EACCES',
2024-05-21 13:44:41.278959500    errno: -13,
2024-05-21 13:44:41.278963500    syscall: 'listen',
2024-05-21 13:44:41.278967500    address: '::0',
2024-05-21 13:44:41.278970500    port: 25
2024-05-21 13:44:41.278974500  }
```

It seems issue is opening port after droping priviledges (missing await?). I am going to investigate it now

--- a/Changes.md
+++ b/Changes.md
@@ -96,6 +96,7 @@
 - fix(outbound): replace empty Message-ID header #3288
 - fix(outbound): don't send SNI servername when connecting to an IP
 - fix(outbound): chown queue dir after creation #3291
+- fix(server): async endpoint.bind() and await in server.js #3366
 
 ### [3.0.3] - 2024-02-07
 

--- a/endpoint.js
+++ b/endpoint.js
@@ -1,7 +1,7 @@
 'use strict';
 // Socket address parser/formatter and server binding helper
 
-const fs = require('node:fs');
+const fs = require('node:fs/promises');
 const sockaddr = require('sockaddr');
 
 module.exports = function endpoint (addr, defaultPort) {
@@ -52,16 +52,16 @@ class Endpoint {
         const mode = this.mode ? parseInt(this.mode, 8) : false;
         if (this.path) {
             opts.path = this.path;
-            if (fs.existsSync(opts.path)) fs.unlinkSync(opts.path);
+            if (await fs.exists(opts.path)) await fs.unlink(opts.path);
         } else {
             opts.host = this.host;
             opts.port = this.port;
         }
         
-        return await new Promise((resolve, reject) => {
-            server.listen(opts, (err) => {
+        return new Promise((resolve, reject) => {
+            server.listen(opts, async (err) => {
                 if(err) return reject(err);
-                if (mode) fs.chmodSync(opts.path, mode);
+                if (mode) await fs.chmod(opts.path, mode);
                 resolve()
             });
         });

--- a/endpoint.js
+++ b/endpoint.js
@@ -52,7 +52,7 @@ class Endpoint {
         const mode = this.mode ? parseInt(this.mode, 8) : false;
         if (this.path) {
             opts.path = this.path;
-            if (fs.existsSync(path)) fs.unlinkSync(path);
+            if (fs.existsSync(opts.path)) fs.unlinkSync(opts.path);
         } else {
             opts.host = this.host;
             opts.port = this.port;

--- a/endpoint.js
+++ b/endpoint.js
@@ -46,21 +46,24 @@ class Endpoint {
     }
 
     // Make server listen on this endpoint, w/optional options
-    bind (server, opts) {
-        let done;
-        opts = Object.assign({}, opts || {});
+    async bind (server, opts) {
+        opts = {...opts};
+
+        const mode = this.mode ? parseInt(this.mode, 8) : false;
         if (this.path) {
-            const path = opts.path = this.path;
-            const mode = this.mode ? parseInt(this.mode, 8) : false;
-            if (mode) {
-                done = () => fs.chmodSync(path, mode);
-            }
+            opts.path = this.path;
             if (fs.existsSync(path)) fs.unlinkSync(path);
-        }
-        else {
+        } else {
             opts.host = this.host;
             opts.port = this.port;
         }
-        server.listen(opts, done);
+        
+        return await new Promise((resolve, reject) => {
+            server.listen(opts, (err) => {
+                if(err) return reject(err);
+                if (mode) fs.chmodSync(opts.path, mode);
+                resolve()
+            });
+        });
     }
 }

--- a/endpoint.js
+++ b/endpoint.js
@@ -52,7 +52,7 @@ class Endpoint {
         const mode = this.mode ? parseInt(this.mode, 8) : false;
         if (this.path) {
             opts.path = this.path;
-            if (await fs.exists(opts.path)) await fs.unlink(opts.path);
+            await fs.rm(this.path, { force: true }); // errors are ignored when force is true
         } else {
             opts.host = this.host;
             opts.port = this.port;

--- a/server.js
+++ b/server.js
@@ -505,7 +505,7 @@ Server.setup_http_listeners = async () => {
             Server.logerror(e);
         })
 
-        ep.bind(Server.http.server, {backlog: 0});
+        await ep.bind(Server.http.server, {backlog: 0});
     }
 
     Server.plugins.run_hooks('init_http', Server);

--- a/server.js
+++ b/server.js
@@ -446,7 +446,7 @@ Server.setup_smtp_listeners = async (plugins2, type, inactivity_timeout) => {
                 Server.logerror(e)
             })
 
-        ep.bind(server, {backlog: 0});
+        await ep.bind(server, {backlog: 0});
     }
 
     if (errors.length) {
@@ -459,7 +459,7 @@ Server.setup_smtp_listeners = async (plugins2, type, inactivity_timeout) => {
     plugins2.run_hooks(`init_${type}`, Server);
 }
 
-Server.setup_http_listeners = () => {
+Server.setup_http_listeners = async () => {
     if (!Server.http?.cfg?.listen) return;
 
     const listeners = Server.get_listen_addrs(Server.http.cfg, 80);

--- a/test/endpoint.js
+++ b/test/endpoint.js
@@ -45,16 +45,16 @@ describe('endpoint', () => {
             }
 
             this.mockfs = {
-                existsSync (path, ...args) {
-                    log.push(['existsSync', path, ...args]);
+                exists (path, ...args) {
+                    log.push(['exists', path, ...args]);
                     return ('undefined' !== typeof modes[path]);
                 },
-                chmodSync (path, mode, ...args) {
-                    log.push(['chmodSync', path, mode, ...args]);
+                chmod (path, mode, ...args) {
+                    log.push(['chmod', path, mode, ...args]);
                     modes[path] = mode;
                 },
-                unlinkSync (path, ...args) {
-                    log.push(['unlinkSync', path, ...args]);
+                unlink (path, ...args) {
+                    log.push(['unlink', path, ...args]);
                     if ('undefined' !== typeof modes[path]) {
                         delete modes[path];
                     }
@@ -64,13 +64,13 @@ describe('endpoint', () => {
                 },
             };
 
-            mock('node:fs', this.mockfs);
+            mock('node:fs/promises', this.mockfs);
             this.endpoint = mock.reRequire('../endpoint');
             done();
         })
 
         afterEach((done) => {
-            mock.stop('node:fs');
+            mock.stop('node:fs/promises');
             done();
         })
 
@@ -86,7 +86,7 @@ describe('endpoint', () => {
             await this.endpoint('/foo/bar.sock').bind(this.server, {readableAll:true});
             assert.deepEqual(
                 this.log, [
-                    ['existsSync', '/foo/bar.sock'],
+                    ['exists', '/foo/bar.sock'],
                     ['listen', {path: '/foo/bar.sock', readableAll: true}],
                 ]);
         })
@@ -96,8 +96,8 @@ describe('endpoint', () => {
             await this.endpoint('/foo/bar.sock').bind(this.server);
             assert.deepEqual(
                 this.log, [
-                    ['existsSync', '/foo/bar.sock'],
-                    ['unlinkSync', '/foo/bar.sock'],
+                    ['exists', '/foo/bar.sock'],
+                    ['unlink', '/foo/bar.sock'],
                     ['listen', {path: '/foo/bar.sock'}],
                 ]);
         })
@@ -106,9 +106,9 @@ describe('endpoint', () => {
             await this.endpoint('/foo/bar.sock:764').bind(this.server);
             assert.deepEqual(
                 this.log, [
-                    ['existsSync', '/foo/bar.sock'],
+                    ['exists', '/foo/bar.sock'],
                     ['listen', {path: '/foo/bar.sock'}],
-                    ['chmodSync', '/foo/bar.sock', 0o764],
+                    ['chmod', '/foo/bar.sock', 0o764],
                 ]);
         })
     })

--- a/test/endpoint.js
+++ b/test/endpoint.js
@@ -74,16 +74,16 @@ describe('endpoint', () => {
             done();
         })
 
-        it('IP socket', () => {
-            this.endpoint('10.0.0.3:42').bind(this.server, {backlog:19});
+        it('IP socket', async () => {
+            await this.endpoint('10.0.a0.3:42').bind(this.server, {backlog:19});
             assert.deepEqual(
                 this.log, [
                     ['listen', {host: '10.0.0.3', port: 42, backlog: 19}],
                 ]);
         })
 
-        it('Unix socket', () => {
-            this.endpoint('/foo/bar.sock').bind(this.server, {readableAll:true});
+        it('Unix socket', async () => {
+            await this.endpoint('/foo/bar.sock').bind(this.server, {readableAll:true});
             assert.deepEqual(
                 this.log, [
                     ['existsSync', '/foo/bar.sock'],
@@ -91,9 +91,9 @@ describe('endpoint', () => {
                 ]);
         })
 
-        it('Unix socket (pre-existing)', () => {
+        it('Unix socket (pre-existing)', async () => {
             this.modes['/foo/bar.sock'] = 0o755;
-            this.endpoint('/foo/bar.sock').bind(this.server);
+            await this.endpoint('/foo/bar.sock').bind(this.server);
             assert.deepEqual(
                 this.log, [
                     ['existsSync', '/foo/bar.sock'],
@@ -102,8 +102,8 @@ describe('endpoint', () => {
                 ]);
         })
 
-        it('Unix socket w/mode', () => {
-            this.endpoint('/foo/bar.sock:764').bind(this.server);
+        it('Unix socket w/mode', async () => {
+            await this.endpoint('/foo/bar.sock:764').bind(this.server);
             assert.deepEqual(
                 this.log, [
                     ['existsSync', '/foo/bar.sock'],

--- a/test/endpoint.js
+++ b/test/endpoint.js
@@ -45,22 +45,12 @@ describe('endpoint', () => {
             }
 
             this.mockfs = {
-                exists (path, ...args) {
-                    log.push(['exists', path, ...args]);
-                    return ('undefined' !== typeof modes[path]);
-                },
                 chmod (path, mode, ...args) {
                     log.push(['chmod', path, mode, ...args]);
                     modes[path] = mode;
                 },
-                unlink (path, ...args) {
-                    log.push(['unlink', path, ...args]);
-                    if ('undefined' !== typeof modes[path]) {
-                        delete modes[path];
-                    }
-                    else {
-                        log.push(['unlink without existing socket']);
-                    }
+                rm (path, ...args) {
+                    log.push(['rm', path, ...args]);
                 },
             };
 
@@ -86,19 +76,8 @@ describe('endpoint', () => {
             await this.endpoint('/foo/bar.sock').bind(this.server, {readableAll:true});
             assert.deepEqual(
                 this.log, [
-                    ['exists', '/foo/bar.sock'],
+                    ['rm', '/foo/bar.sock', {force:true}],
                     ['listen', {path: '/foo/bar.sock', readableAll: true}],
-                ]);
-        })
-
-        it('Unix socket (pre-existing)', async () => {
-            this.modes['/foo/bar.sock'] = 0o755;
-            await this.endpoint('/foo/bar.sock').bind(this.server);
-            assert.deepEqual(
-                this.log, [
-                    ['exists', '/foo/bar.sock'],
-                    ['unlink', '/foo/bar.sock'],
-                    ['listen', {path: '/foo/bar.sock'}],
                 ]);
         })
 
@@ -106,7 +85,7 @@ describe('endpoint', () => {
             await this.endpoint('/foo/bar.sock:764').bind(this.server);
             assert.deepEqual(
                 this.log, [
-                    ['exists', '/foo/bar.sock'],
+                    ['rm', '/foo/bar.sock', {force:true}],
                     ['listen', {path: '/foo/bar.sock'}],
                     ['chmod', '/foo/bar.sock', 0o764],
                 ]);

--- a/test/endpoint.js
+++ b/test/endpoint.js
@@ -75,7 +75,7 @@ describe('endpoint', () => {
         })
 
         it('IP socket', async () => {
-            await this.endpoint('10.0.a0.3:42').bind(this.server, {backlog:19});
+            await this.endpoint('10.0.0.3:42').bind(this.server, {backlog:19});
             assert.deepEqual(
                 this.log, [
                     ['listen', {host: '10.0.0.3', port: 42, backlog: 19}],

